src/mod_stream_management.erl

```diff
@@ -59,19 +73,48 @@ sm() ->
     #xmlel{name = <<"sm">>,
            attrs = [{<<"xmlns">>, ?NS_STREAM_MGNT_3}]}.
 
+-spec remove_smid(Acc, SID, JID, Info, Reason) -> Acc1 when
+      Acc :: mongoose_acc:t(),
+      SID :: ejabberd_sm:sid(),
+      JID :: undefined | jid:jid(),
+      Info :: undefined | [any()],
+      Reason :: undefined | ejabberd_sm:close_reason(),
+      Acc1 :: mongoose_acc:t().
 remove_smid(Acc, SID, _JID, _Info, _Reason) ->
-    case mnesia:dirty_index_read(sm_session, SID, #sm_session.sid) of
-        [] ->
-            ok;
-        [#sm_session{} = SMSession] ->
-            mnesia:sync_dirty(fun mnesia:delete_object/1, [SMSession])
-    end,
-    Acc.
+    do_remove_smid(Acc, SID).
 
 -spec session_cleanup(Acc :: map(), LUser :: jid:luser(), LServer :: jid:lserver(),
                       LResource :: jid:lresource(), SID :: ejabberd_sm:sid()) -> any().
 session_cleanup(Acc, _LUser, _LServer, _LResource, SID) ->
-    remove_smid(Acc, SID, undefined, undefined, undefined).
+    Acc1 = do_remove_smid(Acc, SID),
+    case mongoose_acc:get(stream_mgmt, smid, Acc1) of
+        {error, smid_not_found} -> ok;
+        {ok, SMID} -> remove_stale_smid_h(SMID)
```

Comment:
The aim of `session_cleanup` is to remove session entries bound to a MIM node that died/crashed/shutdown. Is it even possible to have stale SMID h for a session that is currently alive?